(FACT-1177) Add swap facts to OpenBSD memory resolver

jasperla opened the pull request:
These bits were initially missing, so now this resolver is complete.

MikaelSmith commented on lib/src/facts/openbsd/memory_resolver.cc:
```diff
@@ -37,6 +37,38 @@ namespace facter { namespace facts { namespace openbsd {
+        // NB: swapctl(2) for SWAP_NSWAP cannot fail
+        nswaps = swapctl(SWAP_NSWAP, 0, 0);
+
+        if ((swapdev = (struct swapent *)calloc(nswaps, sizeof(*swapdev))) == NULL)
```
This seems to be leaked memory. To avoid this, we've tended to use [RAII](https://en.wikipedia.org/wiki/Resource_Acquisition_Is_Initialization), as in https://github.com/puppetlabs/facter/blob/master/lib/inc/internal/util/posix/scoped_addrinfo.hpp.

MikaelSmith replied:
Also, please prefer `{ ... }`, even for single-line `if` statements.

MikaelSmith commented on lib/src/facts/openbsd/memory_resolver.cc:
```diff
@@ -37,6 +37,38 @@ namespace facter { namespace facts { namespace openbsd {
+        len = sizeof(encrypted);
+
+        if (sysctl(swap_encrypted_mib, 3, &encrypted, &len, nullptr, 0) == -1) {
+                LOG_DEBUG("sysctl failed: %1% (%2%): encrypted swap fact not available.", strerror(errno), errno);
```
Please fix indentation here.

MikaelSmith commented on lib/src/facts/openbsd/memory_resolver.cc:
```diff
@@ -37,6 +37,38 @@ namespace facter { namespace facts { namespace openbsd {
+        if (swapctl(SWAP_STATS, swapdev, nswaps) == -1)
+             LOG_DEBUG("swapctl: SWAP_STATS failed: %1% (%2%)", strerror(errno), errno);
+
+        for (s = 0; s < nswaps; s++) {
```
With a vector, you can then use range-for loops:
```
for (auto &&swap : swapdev) {
    if (swap.se_flags & SWF_ENABLE) {
        result.swap_total += swap.se_nblks * DEV_BSIZE;
        swap_used += swap.se_inuse * DEV_BSIZE;
    }
}
```

MikaelSmith commented:
AppVeyor failure is a red-herring, should pass when you push fixes.

jasperla commented:
I've fixed the spacing/brackets issues, and addressed the memleak by actually freeing `swapdev`. I've left out the vector bits as I don't fully understand that part.

I naively tried something along these lines, which blew up with 2 screens full of `g++` output:

```
        int nswaps, s;
        uint64_t swap_used = 0;

        // NB: swapctl(2) for SWAP_NSWAP cannot fail
        nswaps = swapctl(SWAP_NSWAP, 0, 0);

        vector<struct swapent> swapdev(nswaps);
        if ((swapdev = (struct swapent *)calloc(nswaps, sizeof(swapdev.data()))) == NULL) {
            LOG_DEBUG("cannot allocate swapdev.");
            return result;
	}

        if (swapctl(SWAP_STATS, swapdev, nswaps) == -1) {
             LOG_DEBUG("swapctl: SWAP_STATS failed: %1% (%2%)", strerror(errno), errno);
	     return result;
	}

	for (auto &&swap : swapdev) {
	  if (swap.se_flags & SWF_ENABLE) {
	      result.swap_total += swap.se_nblks * DEV_BSIZE;
	      swap_used += swap.se_inuse * DEV_BSIZE;
	  }
        }
```

MikaelSmith commented:
You wouldn't want to calloc swapdev.data(), it's already been allocated and initialized, and you would want to pass swapdev.data() to swapctl.

Does https://github.com/puppetlabs/facter/compare/master...MikaelSmith:jasperla-openbsd/swap work?

jasperla commented:
Ah, I see what you mean. I'm still trying to wrap my head around C++. I've updated the patch with all your suggestions now. Thanks.

MikaelSmith commented on lib/src/facts/openbsd/memory_resolver.cc:
```diff
@@ -19,24 +19,54 @@ namespace facter { namespace facts { namespace openbsd {
-            // Should we account for the buffer cache?
-            result.mem_total = static_cast<u_int64_t>(uvmexp.npages) << uvmexp.pageshift;
-            result.mem_free = static_cast<u_int64_t>(uvmexp.free) << uvmexp.pageshift;
+                int uvmexp_mib[] = { CTL_VM, VM_UVMEXP };
```
Not sure what's going on with indentation here.

MikaelSmith commented on lib/src/facts/openbsd/memory_resolver.cc:
```diff
@@ -19,24 +19,54 @@ namespace facter { namespace facts { namespace openbsd {
+
+        uint64_t swap_used = 0;
+        for (auto &&swap : swapdev) {
+          if (swap.se_flags & SWF_ENABLE) {
```
Also indentation here.

jasperla commented:
Quirky indentation's fixed too now.

MikaelSmith commented:
Thanks, looks good now.
